Add ability to display depth buffer in editor 3d preview.

Joxno opened the pull request:
This pull request implements the ability to quickly display the depth buffer in editor 3d preview and fixes a bug in the copy.glsl shader related to how linear depth is calculated.

![image](https://user-images.githubusercontent.com/9378113/197308343-243fbcf4-0e8b-4721-8520-0f86aee129ed.png)

![image](https://user-images.githubusercontent.com/9378113/197308393-8c885e99-ad10-4b2e-9b16-10b957bca310.png)

Commit:
Fixed bug with incorrect linear depth calculation. Added mode for reversing depth values.

Commit:
Implemented ability to display depth buffer in editor 3d preview.

clayjohn commented on servers/rendering/renderer_rd/shaders/effects/copy.glsl:
```diff
@@ -241,12 +241,18 @@ void main() {
 	float depth = texelFetch(source_color, pos + params.section.xy, 0).r;
-	depth = depth * 2.0 - 1.0;
-	depth = 2.0 * params.camera_z_near * params.camera_z_far / (params.camera_z_far + params.camera_z_near - depth * (params.camera_z_far - params.camera_z_near));
+
+	depth = params.camera_z_near * params.camera_z_far /
+			(params.camera_z_far + depth * (params.camera_z_near - params.camera_z_far));
```
Could you explain in more detail what is wrong with this way of linearizing depth? This is the same method used throughout the engine, so if it changes here it should change there as well

Joxno replied:
I think I was a bit quick in editing that shader, I've restored line 248 in a commit I'll be pushing.
In terms of line 247, if I have understood the maths correctly that sort of conversion should not be needed in Vulkan compared to OpenGL, is it correct that it is no longer needed?

clayjohn replied:
No, it is still needed. We aren't dealing with NDC values here, we are dealing with the contents of the depth buffer which are still values in the 0-1 range regardless of using Vulkan or OpenGL. 

If you look at other parts of the codebase, they all use the same function to convert depth buffer depth into linear depth. I would stick with using the same code here as elsewhere so that everything matches.

clayjohn commented on servers/rendering/renderer_rd/shaders/effects/copy.glsl:
```diff
@@ -241,12 +241,18 @@ void main() {
+	depth = params.camera_z_near * params.camera_z_far /
+			(params.camera_z_far + depth * (params.camera_z_near - params.camera_z_far));
+
+#ifdef MODE_LINEARIZE_DEPTH_COPY_INVERSED
+	vec4 color = vec4(1.0 - (depth / params.camera_z_far));
+#else
```
I wouldn't add a new shader version for this, I would just add a single boolean variable to the push constant and do:
```
vec4 color = vec4(depth / params.camera_z_far);
if (params.inverse_depth) {
    color = 1.0 - color;
}
```

Joxno replied:
Ah, yeah that is definitely a better way of doing it, I'll change it to use a push constant value instead.

clayjohn commented on servers/rendering/renderer_rd/effects/copy_effects.cpp:
```diff
@@ -500,6 +501,47 @@ void CopyEffects::copy_depth_to_rect_and_linearize(RID p_source_rd_texture, RID
 	RD::get_singleton()->compute_list_end();
 }
 
+void CopyEffects::copy_depth_to_rect_and_linearize_reversed(RID p_source_rd_texture, RID p_dest_texture, const Rect2i &p_rect, bool p_flip_y, float p_z_near, float p_z_far) {
```
Similar to my other comment, we don't need a whole different version to reverse the depth. I would instead just add a boolean flag to the push constant and add a parameter to ``copy_depth_to_rect_and_linearize()`` that defaults to false like: ``p_invert = false``

Joxno replied:
I was going back and forth pondering if I wanted to change the original signature or not and ended up just going with the new function. I'll change it to your suggestion instead.

On a side note, there are a few functions in "copy_effects.cpp" that are very similar with the exception for push constants and mode, might be worth a future refactor to maybe make a more generic function that takes in those two things as parameters instead to save the duplication of code.

clayjohn reviewed (commented):
Good idea! I think this will be nice addition and it makes sense seeing as how we already have a display mode for the normal buffer. 

One thing I am unsure about is whether we actually need to invert the depth. I left some comments on implementation assuming we keep the inverted depth code, but to me it seems like it may make more sense for the debug mode to show users something closer to the actual value instead of inverting the range. Did you do it this way because non-inverted depth was too difficult to read?

RedMser commented:
@Joxno You need to run `godot --doctool` to generate the XML documentation for the newly added enum entries.

Commit:
Restored linear depth calculation
Changed mode to push constant instead after feedback.
Changed method of separate copy function to boolean flag instead after feedback.

Joxno commented:
> Good idea! I think this will be nice addition and it makes sense seeing as how we already have a display mode for the normal buffer.
> 
> One thing I am unsure about is whether we actually need to invert the depth. I left some comments on implementation assuming we keep the inverted depth code, but to me it seems like it may make more sense for the debug mode to show users something closer to the actual value instead of inverting the range. Did you do it this way because non-inverted depth was too difficult to read?

Thanks!
Yeah, I can understand the argument to display closer to the truth values for debugging purposes although I think without an inversion it makes the visualisation harder to read.

Here's an example if you do not invert the values:
![image](https://user-images.githubusercontent.com/9378113/197361383-bbec15d0-809a-4d33-8a83-8552a29bfff7.png)

Obviously this is a trivial scene with an infinite background but I think in a more complex scene objects that occupy the middle-ground would be washed out. Possibly making it difficult to debug depth issues.

Maybe having 2 alternatives to choose from, one that displays the values as is, and a second one that is inverted could be a possibility?
"Depth Buffer"
"Depth Buffer (Inv)"

It could also be that I am biased to dark mode haha.

clayjohn commented on doc/classes/Viewport.xml:
```diff
@@ -436,54 +436,56 @@
 		<constant name="DEBUG_DRAW_NORMAL_BUFFER" value="5" enum="DebugDraw">
 		</constant>
-		<constant name="DEBUG_DRAW_VOXEL_GI_ALBEDO" value="6" enum="DebugDraw">
+		<constant name="DEBUG_DRAW_DEPTH_BUFFER" value="6" enum="DebugDraw">
```
This should be filled in

clayjohn commented on servers/rendering/renderer_rd/effects/copy_effects.h:
```diff
@@ -155,6 +155,9 @@ class CopyEffects {
 		uint32_t pad2[2];
 		//SET color
 		float set_color[4];
+		// Depth.
+		bool inverse_depth;
+		uint32_t pad3[3];
```
This should replace one of the pads above (I would replace ``pad``) There is no reason to introduce 3 new pads when there is already so much padding

clayjohn commented:
Discussed in the PR meeting today. Consensus was that this looks mostly good, but both non-inverse and inverse result in far away objects being difficult to see. Lawnjelly helpfully suggested (and the other contributors agreed) that we should use the non-inverted linear depth, but we should set the far plane to a fixed color so that far objects contrast well against it.

I am thinking a dark blue colour from this page would be good https://davidmathlogic.com/colorblind/#%23D81B60-%231E88E5-%23FFC107-%23004D40. There are a few colours suggested here that have good visibility for many different types of colourblindness.

Calinou commented:
@Joxno Could you look into applying clayjohn's suggestions above? The PR shoUld be good to merge afterwards.

AThousandShips commented on doc/classes/RenderingServer.xml:
```diff
@@ -4727,69 +4727,72 @@
 			Normal buffer is drawn instead of regular scene so you can see the per-pixel normals that will be used by post-processing effects.
 		</constant>
-		<constant name="VIEWPORT_DEBUG_DRAW_VOXEL_GI_ALBEDO" value="6" enum="ViewportDebugDraw">
+		<constant name="VIEWPORT_DEBUG_DRAW_DEPTH_BUFFER" value="6" enum="ViewportDebugDraw">
```
Does this have to be added at this position or can it be added at the end? This changes a lot of values that affects extensions and binary compatibility AFAIK so would be nice to be as non-disruptive as possible

Joxno replied:
Moved enums values to the end as to not cause any conflicts.

AThousandShips commented on doc/classes/Viewport.xml:
```diff
@@ -519,54 +519,56 @@
-		<constant name="DEBUG_DRAW_MOTION_VECTORS" value="25" enum="DebugDraw">
+		<constant name="DEBUG_DRAW_MOTION_VECTORS" value="26" enum="DebugDraw">
 		</constant>
 		<constant name="DEBUG_DRAW_INTERNAL_BUFFER" value="26" enum="DebugDraw">
```
This needs to be updated too, you need to run doctool again I think

Joxno replied:
Should be corrected in the latest commit.

AThousandShips commented on servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:
```diff
@@ -776,6 +776,29 @@ void RendererSceneRenderRD::_render_buffers_debug_draw(const RenderDataRD *p_ren
 		copy_effects->copy_to_fb_rect(_render_buffers_get_normal_texture(rb), texture_storage->render_target_get_rd_framebuffer(render_target), Rect2(Vector2(), rtsize), false, false, false, false, RID(), false, false, false, true);
 	}
 
+	if (debug_draw == RS::VIEWPORT_DEBUG_DRAW_DEPTH_BUFFER && _render_buffers_get_depth_texture(p_render_buffers).is_valid()) {
+		Size2 rtsize = texture_storage->render_target_get_size(render_target);
+
+		RD::TextureSamples texture_samples = RD::TEXTURE_SAMPLES_1;
+		RD::DataFormat format = _render_buffers_get_color_format();
+		uint32_t usage_bits = RD::TEXTURE_USAGE_SAMPLING_BIT | RD::TEXTURE_USAGE_CAN_COPY_TO_BIT | RD::TEXTURE_USAGE_STORAGE_BIT | RD::TEXTURE_USAGE_COLOR_ATTACHMENT_BIT;
+
+		RID debug_depth_texture = p_render_buffers->has_texture(RB_SCOPE_FORWARD_CLUSTERED, SNAME("debug_depth")) ? p_render_buffers->get_texture(RB_SCOPE_FORWARD_CLUSTERED, SNAME("debug_depth")) : p_render_buffers->create_texture(RB_SCOPE_FORWARD_CLUSTERED, SNAME("debug_depth"), format, usage_bits, texture_samples);
+
+		copy_effects->copy_depth_to_rect_and_linearize(
+				_render_buffers_get_depth_texture(p_render_buffers),
```
```suggestion
	if (debug_draw == RS::VIEWPORT_DEBUG_DRAW_DEPTH_BUFFER && _render_buffers_get_depth_texture(rb).is_valid()) {
		Size2 rtsize = texture_storage->render_target_get_size(render_target);

		RD::TextureSamples texture_samples = RD::TEXTURE_SAMPLES_1;
		RD::DataFormat format = _render_buffers_get_color_format();
		uint32_t usage_bits = RD::TEXTURE_USAGE_SAMPLING_BIT | RD::TEXTURE_USAGE_CAN_COPY_TO_BIT | RD::TEXTURE_USAGE_STORAGE_BIT | RD::TEXTURE_USAGE_COLOR_ATTACHMENT_BIT;

		RID debug_depth_texture = rb->has_texture(RB_SCOPE_FORWARD_CLUSTERED, SNAME("debug_depth")) ? rb->get_texture(RB_SCOPE_FORWARD_CLUSTERED, SNAME("debug_depth")) : rb->create_texture(RB_SCOPE_FORWARD_CLUSTERED, SNAME("debug_depth"), format, usage_bits, texture_samples);

		copy_effects->copy_depth_to_rect_and_linearize(
				_render_buffers_get_depth_texture(rb),
```

Joxno replied:
Apologies, seems I accidentally pushed to origin as I was rebasing, should be fixed in the latest commit.

Joxno committed:
Removed inverse depth functionality. Fixed issues with depth linear calculations. Added background colour at max depth. Corrected and moved documentation.

Joxno committed:
Fixed style issue in copy shader.

Joxno committed:
Fixed issue with debug depth viewport enum binding having wrong order.

Calinou commented:
Tested locally, there seems to be a mismatch between the UI in the Perspective menu and the actual debug draw mode you get:

https://github.com/godotengine/godot/assets/180032/31d4dd12-db45-4cea-acc1-84ac3d1d8605

**Testing project:** [test_shadowmask.zip](https://github.com/godotengine/godot/files/14288677/test_shadowmask.zip)

Joxno committed:
Fixed issue with wrong order in display_options array.

Joxno commented:
Ah, I made a complete blunder when I reordered the debug_draw_modes array but forgot to take into account the order in display_options as well. Should be resolved in the latest commit.

![image](https://github.com/godotengine/godot/assets/9378113/01f66e2c-0f9d-4fc9-935b-458eaaee5242)

Calinou commented on servers/rendering/renderer_rd/shaders/effects/copy.glsl:
```diff
@@ -245,9 +247,22 @@ void main() {
+	vec4 color = vec4(depth_linear_normalized, depth_linear_normalized, depth_linear_normalized, 1.0);
+
+	if (depth >= z_far - 0.01) {
+		color = vec4(0.0, 0.2705, 0.5490, 1.0);
```
Out of curiosity, was this shade of blue chosen arbitrarily or does it have special meaning?

Calinou reviewed (commented):
Tested locally, it works as expected in Forward+. I can't get this debug draw mode to show up when using the Mobile rendering method though, even though it should be able to work according to what I see in the diff.
It would be nice to get it working in Compatibility too, but this can be left for a future PR.

I noticed there's quite a lot of banding in the debug view:

![image](https://github.com/godotengine/godot/assets/180032/4e51bca5-019f-4abf-8fdb-8b6e233de058)

Even with camera near set to `1.0` and far set to `25.0` in the editor view settings, I still get a fair amount of banding:

![image](https://github.com/godotengine/godot/assets/180032/1231bd1a-62d2-4686-ac09-308143863d06)

It appears the stepping is actually correct (there are no shades of gray "missed" between the bands), but since the depth is displayed in a non-linear manner, the banding in near-black shades gets pretty obvious. This may be less noticeable depending on viewing conditions or the display you're using (I'm using a LG C2 42" at 120 nits brightness in SDR).

PS: This PR will likely need changes if https://github.com/godotengine/godot/pull/88328 is merged before this one.

Joxno committed:
Implemented stubbed mobile function for retrieving depth texture.

Joxno committed:
Changed the hardcoded zNear, zFar values to actual scene camera projection values.

aaronfranke commented:
@Joxno Can you rebase this PR? It would be nice to have this feature in the engine.

mooflu commented:
Ignore commit reference above. 
See this one instead: https://github.com/godotengine/godot/commit/6a3ebdf08786e2ebdcf19a1c01b3eac6f15df606

Contains most of the original except camera_z_far_scale and some of the shader changes.
Changed MODE_LINEARIZE_DEPTH_COPY with inverted depth and special color when close to far z.
The original had special ortho handling, but I don't think it was working. It used an ifdef and at least now it would need a flag check `bool(params.flags & FLAG_USE_ORTHOGONAL_PROJECTION)` but maybe it was different back then...

I can do a new PR for this or maybe someone with magic powers can apply it to this one. 

<img width="804" height="605" alt="d1-cam-persp" src="https://github.com/user-attachments/assets/b910e132-d5da-41c0-8c69-329e6d7f0147" />
